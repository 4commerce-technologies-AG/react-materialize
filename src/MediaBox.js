import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import idgen from './idgen';

const MediaBox = ({ id, options, children, className, caption, ...props }) => {
  useEffect(() => {
    const instance = M.Materialbox.init(document.getElementById(id), options);

    return () => {
      if (instance) {
        instance.destroy();
      }
    };
  }, [id, options]);

  return React.cloneElement(children, {
    id: id,
    className: cx('materialboxed', className),
    'data-caption': caption,
    ...props
  });
};

MediaBox.propTypes = {
  children: PropTypes.node.isRequired,
  /*
   * override id
   * @default idgen()
   */
  id: PropTypes.string,
  className: PropTypes.string,
  /**
   * The caption shown below the image when opened
   */
  caption: PropTypes.string,
  options: PropTypes.shape({
    /**
     * Transition in duration in milliseconds.
     */
    inDuration: PropTypes.number,
    /**
     * Transition out duration in milliseconds.
     */
    outDuration: PropTypes.number,
    /**
     * Callback function called before materialbox is opened.
     */
    onOpenStart: PropTypes.func,
    /**
     * Callback function called after materialbox is opened.
     */
    onOpenEnd: PropTypes.func,
    /**
     * Callback function called before materialbox is closed.
     */
    onCloseStart: PropTypes.func,
    /**
     * Callback function called after materialbox is closed.
     */
    onCloseEnd: PropTypes.func
  })
};

MediaBox.defaultProps = {
  id: `MediaBox_${idgen()}`,
  options: {
    inDuration: 275,
    outDuration: 200,
    onOpenStart: null,
    onOpenEnd: null,
    onCloseStart: null,
    onCloseEnd: null
  }
};

export default MediaBox;
