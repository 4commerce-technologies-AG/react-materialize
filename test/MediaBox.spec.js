import React from 'react';
import { shallow, mount } from 'enzyme';
import MediaBox from '../src/MediaBox';
import mocker from './helper/new-mocker';

describe('<MediaBox />', () => {
  let wrapper;
  const materialboxInitMock = jest.fn();
  const materialboxInstanceDestroyMock = jest.fn();
  const materialboxMock = {
    init: (el, options) => {
      materialboxInitMock(options);
      return {
        destroy: materialboxInstanceDestroyMock
      };
    }
  };
  const restore = mocker('Materialbox', materialboxMock);

  afterAll(() => {
    restore();
  });

  test('renders', () => {
    wrapper = shallow(
      <MediaBox
        className="more"
        image={<img src="image.jpg" alt="" />}
        caption="A demo media box1"
        width="650"
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  describe('initialises', () => {
    beforeEach(() => {
      materialboxInitMock.mockClear();
      materialboxInstanceDestroyMock.mockClear();
    });

    test('calls Materialbox', () => {
      mount(<MediaBox image={<img src="image.jpg" alt="" />} />);

      expect(materialboxInitMock).toHaveBeenCalledTimes(1);
    });

    test('calls Materialbox with options', () => {
      const options = {
        inDuration: 325,
        outDuration: 250
      };

      mount(
        <MediaBox image={<img src="image.jpg" alt="" />} options={options} />
      );

      expect(materialboxInitMock).toHaveBeenCalledWith(options);
    });
  });
});
