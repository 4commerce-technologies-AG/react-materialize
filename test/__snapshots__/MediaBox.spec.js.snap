// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MediaBox /> renders 1`] = `
<img
  alt=""
  className="materialboxed more"
  data-caption="A demo media box1"
  id="MediaBox_mockId"
  src="image.jpg"
  width="650"
/>
`;
